src/media/class-wpml-elementor-media-nodes-iterator.php

```diff
@@ -19,6 +19,9 @@ public function __construct( WPML_Elementor_Media_Node_Provider $node_provider )
 	public function translate( $data_array, $lang, $source_lang ) {
 		foreach ( $data_array as &$node ) {
 			if ( $this->is_parent_node( $node ) ) {
+				if ( $this->is_background_image_set( $node ) ) {
+					$node = $this->translate_node_background_image( $node, $source_lang, $lang );
```

Comment:
@GeorgeBotsev, when we started this project we were only translating strings (in this class).

Then, we started supporting media translation in a separate layer. See https://github.com/OnTheGoSystems/wpml-page-builders-elementor/blob/master/src/media/class-wpml-elementor-media-nodes-iterator.php#L3.

Check if you can insert your code in `WPML_Elementor_Media_Nodes_Iterator` or maybe in the base module class `WPML_Elementor_Media_Nodes_Iterator`.

Reply:
If I understand the problem correctly, we should convert the background image on any king of element. So maybe we need a generic module extending `WPML_Elementor_Media_Node` (e.g. `WPML_Elementor_All_Widget_Media_Node`) that we could get from the `WPML_Elementor_Media_Node_Provider`.

Then:

```php
	public function translate( $data_array, $lang, $source_lang ) {
		foreach ( $data_array as &$node ) {
			if ( $this->is_parent_node( $node ) ) {
				$node['elements'] = $this->translate( $node['elements'], $lang, $source_lang );
			} elseif ( $this->is_valid_media_node( $node ) ) {
				$node = $this->translate_node( $node, $lang, $source_lang );
			}

			// => Here we'll use this "all widget" media node - not sure the naming is correct though
			$node = $this->node_provider->get( 'all_widgets' )->translate( $node_data['settings'], $lang, $source_lang );
		}

		return $data_array;
	}
```

Reply:
> Check if you can insert your code in `WPML_Elementor_Media_Nodes_Iterator`
I am a bit puzzled.
My code is in  `WPML_Elementor_Media_Nodes_Iterator` class.

Reply:
I changed the code - moved it in a separate class.
Let me know what you think @strategio